### PR TITLE
Fix two memory leaks when transcoding fails

### PR DESCRIPTION
**Pull request by elv-reza**
**issue:** https://github.com/eluv-io/avpipe/issues/53
- Fix a big memory leak (> 500 MB) when avpipe transcoding fails due to an audio index pointing to non-audio stream
- Fix another memory leak when transcoding fails and there are some packets in the input channels.
- Avoid transcoding if one of the audio indexes conflicts with another non-audio stream.

### CODE REVIEW

File: libavpipe/src/avpipe_xc.c [hunk 1]
```diff
@@ -390,6 +390,23 @@ decode_interrupt_cb(
     return decoder_ctx->cancelled;
 }
 
+static int
+check_stream_index(
+    xcparams_t *params,
+    coderctx_t *decoder_context)
+{
+    for (int i = 0; i < decoder_context->format_context->nb_streams && i < MAX_STREAMS; i++) {
```
Comment (elv-nate) [inline libavpipe/src/avpipe_xc.c:398]: This for loop is unnecessary, as you don't actually use `i`.

File: libavpipe/src/avpipe_xc.c [hunk 2]
```diff
@@ -4395,6 +4426,13 @@ check_params(
         params->n_audio = 1;
     }
 
+    for (int i=0; i<params->n_audio; i++) {
```
Comment (elv-nate) [inline libavpipe/src/avpipe_xc.c:4429]: I thought when we discussed things earlier, it was reasonable for multiple audios to share the same index if the operation wanted multiple copies of the same audio track? Is this safe for the `AudioPan` or `AudioMerge` usecases? Just want to double check.

File: libavpipe/src/avpipe_xc.c (hunk 1 above)
Comment (elv-reza) [reply libavpipe/src/avpipe_xc.c:398]: ooops!

File: libavpipe/src/avpipe_xc.c (hunk 2 above)
Comment (elv-reza) [reply libavpipe/src/avpipe_xc.c:4429]: No, as I discussed we want to have audio join/merge to use the same audio index. This is not implemented yet, and we can open up this check when we (I) implement this new feature.
Right now, I just avoid having multiple audio so nobody makes a mistake in setting audio indexes.

Comment (elv-nate) [reply libavpipe/src/avpipe_xc.c:4427]: Ah, that makes perfect sense, yes.

**Review by elv-nate — APPROVED** [review]
LGTM

**Review by elv-serban — APPROVED** [review]
Looks good!

File: libavpipe/src/avpipe_xc.c [hunk 3]
```diff
@@ -390,6 +390,21 @@ decode_interrupt_cb(
...
+static int
+check_stream_index(
+    xcparams_t *params,
+    coderctx_t *decoder_context)
+{
+    if (selected_audio_index(params, decoder_context->video_stream_index) >= 0)
+        return eav_param;
+    if (selected_audio_index(params, decoder_context->data_scte35_stream_index) >= 0)
+        return eav_param;
+    if (selected_audio_index(params, decoder_context->data_stream_index) >= 0)
```
[the remark this answers is not in the document]
Comment (elv-reza) [reply libavpipe/src/avpipe_xc.c:402]: assert is good for detecting logical errors in the code. In this case, we have a logical error in the input (by user), and I report this error to the caller (return eav_param error).

The above code, checks to become sure video/data streams are not in the audio indexes by the user.

[the remark this answers is not in the document]
Comment (elv-serban) [reply libavpipe/src/avpipe_xc.c:402]: Sorry I didn't mean "assert" as in `assert()` 
I meant "make sure"

**Pull request merged**